[PyTorch] Add VariableVersion&& overload for TensorImpl::shallow_copy_and_detach

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48681 [PyTorch] Add VariableVersion&& overload for TensorImpl::shallow_copy_and_detach**
* #48680 [PyTorch] Move TensorImpl::shallow_copy_and_detach to .cpp file

This should reduce reference counting traffic when creating views.

The code duplication here is unfortunate and I'm open to suggestions on how to reduce it. It's especially regrettable that we create a footgun for subclasses of TensorImpl: they can accidentally override only one of the two overloads and get confusing behavior.

Differential Revision: [D25259741](https://our.internmc.facebook.com/intern/diff/D25259741/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25259741/)!

swolchok committed:
[PyTorch] Add VariableVersion&& overload for TensorImpl::shallow_copy_and_detach

This should reduce reference counting traffic when creating views.

The code duplication here is unfortunate and I'm open to suggestions on how to reduce it. It's especially regrettable that we create a footgun for subclasses of TensorImpl: they can accidentally override only one of the two overloads and get confusing behavior.

Differential Revision: [D25259741](https://our.internmc.facebook.com/intern/diff/D25259741/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25259741/)!

[ghstack-poisoned]

ezyang commented:
My best idea for deduplicating is to factor shallow_copy_and_detach further: we have a virtualized `_shallow_clone` that is just

```
    auto impl = c10::make_intrusive<QTensorImpl>(
        Storage(storage()), key_set(), data_type_, quantizer_);
```

and then everything else is a nonvirtual method that calls `_shallow_clone`. Seems like it should work, but I only audited the sites that were modified in this PR.

ezyang reviewed (approved):
Going to approve this anyway.

swolchok committed:
Update on "[PyTorch] Add VariableVersion&& overload for TensorImpl::shallow_copy_and_detach"


This should reduce reference counting traffic when creating views.

The code duplication here is unfortunate and I'm open to suggestions on how to reduce it. It's especially regrettable that we create a footgun for subclasses of TensorImpl: they can accidentally override only one of the two overloads and get confusing behavior.

Differential Revision: [D25259741](https://our.internmc.facebook.com/intern/diff/D25259741/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25259741/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add VariableVersion&& overload for TensorImpl::shallow_copy_and_detach"


This should reduce reference counting traffic when creating views.

The code duplication here is unfortunate and I'm open to suggestions on how to reduce it. It's especially regrettable that we create a footgun for subclasses of TensorImpl: they can accidentally override only one of the two overloads and get confusing behavior.

Differential Revision: [D25259741](https://our.internmc.facebook.com/intern/diff/D25259741/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25259741/)!

[ghstack-poisoned]
